Much Anticipated Camp Docs (#158)
LLNL/camp: docs/sphinx/user_guide/using_camp.rst

.. _using_camp-label: 

**********
Using Camp
**********

Since Camp is a metaprogramming library that is primarily header-only, it is hard
to find isolated code examples that show Camp's capability. Instead, this page
shows several real application examples of how Camp is being used currently.

Camp Used in Umpire
===================

> Umpire is a resource management library that allows the discovery, provision, and management of memory on machines 

Reviewer: ```suggestion
`Umpire <https://github.com/LLNL/Umpire>`_ is a resource management library that allows the discovery, provision, and management of memory on machines 
```

with multiple memory devices like NUMA and GPUs. Umpire is also part of the RAJA Portability Suite.

> Umpire's Operations provide an abstract interface to modifying and moving data between Umpire allocators.

Reviewer: ```suggestion
Umpire's Operations provide an abstract interface to modifying and moving data between Umpire Allocators.
```

Operations and Allocators are key concepts in Umpire. Either capitalize both or neither for consistency.

> Camp makes these operations easy to port regardless of the underlying hardware. For example:

Reviewer: This sentence seems to imply that Camp simplifies portability, but the code example is specific to CUDA. Maybe be more explicit in what the code example is meant to illustrate.

Author: Done, thanks.

.. code-block:: cpp

>   camp::resources::EventProxy<camp::resources::Resource> CudaMemsetOperation::apply_async(

Reviewer: Note that keeping the code example in these docs consistent with the Umpire code will require manual care. Please consider linking in the actual code section here rather than cut-and-paste.

Author: The user-facing camp examples in Umpire are mostly just using resources (with the `wait()` and `get_event()` methods) which have already been shown in previous examples. This code snippet shows a bit more. I could look in the umpire tests or benchmarks which are typically a bit more complex than our super high-level dummy examples - stay tuned...

Author: There are a couple examples in Umpire of using literal includes to tell sphinx which lines of code to include into the .rst file... but that's all within Umpire, not linking to some other outside repo like it would be here between camp and umpire. I can try it and see if it works, but the main point in including these examples is just to demonstrate how camp is used in a realistic setting (working code) since Camp has zero examples - Therefore, I'm not sure it really matters if the true source code ends up changing a bit, since it would still be capturing a real, live, working (though out-of-date) camp code example. It's not the best solution, but again, no Camp examples

    if (!device) {
      UMPIRE_ERROR(resource_error,
                 fmt::format("Expected resources::Cuda, got resources::{}", platform_to_string(ctx.get_platform())));
    }
    auto stream = device->get_stream();

    cudaError_t error = ::cudaMemsetAsync(src_ptr, value, length, stream);

    if (error != cudaSuccess) {
      UMPIRE_ERROR(
        runtime_error,
        fmt::format("cudaMemsetAsync( src_ptr = {}, value = {}, length = {}, stream = {}) failed with error: {}",
                    src_ptr, value, length, cudaGetErrorString(error), (void*)stream));
    }

    return camp::resources::EventProxy<camp::resources::Resource>{ctx};
  }

See the full example `here <https://github.com/LLNL/Umpire/blob/5bf5bc182f1e6ee3f6be1d953b68451d3ddc35f5/src/umpire/op/CudaMemsetOperation.cpp>`_.

.. note::

   The new ``ResourceAwarePool`` feature in Umpire will be using both Camp resources and Camp events to
   keep track of the state of different allocations within the pool. Without these Camp features, it would
   be much harder to keep track of the state of the allocations in a portable, hardware-agnostic way.

Camp Used in RAJA
=================

> RAJA is a library of C++ software abstractions, primarily developed at Lawrence Livermore National Laboratory (LLNL), that enables 

Reviewer: ```suggestion
`RAJA <https://github.com/LLNL/RAJA>`_ is a library of C++ software abstractions that enables 
```
The clause about where it is developed will be implicit by linking in the RAJA project home page.

architecture and programming model portability for HPC applications.
One of many examples of using Camp in RAJA is the following code block which helps RAJA determine which backend
is being used in the ``RAJA::Launch`` abstraction.

.. code-block:: cpp

>    // Helper function to retrieve a resource based on the run-time policy - if a device is active

Reviewer: Similar comments about this RAJA internal implementation example as above for the Umpire example.

   #if defined(RAJA_ENABLE_CUDA) || defined(RAJA_ENABLE_HIP) || defined(RAJA_ENABLE_SYCL)
     template<typename T, typename U>
     RAJA::resources::Resource Get_Runtime_Resource(T host_res, U device_res, RAJA::ExecPlace device){
       if(device == RAJA::ExecPlace::DEVICE) {return RAJA::resources::Resource(device_res);}
       else { return RAJA::resources::Resource(host_res); }
       }
   #endif

See the full example `here <https://github.com/LLNL/RAJA/blob/develop/include/RAJA/pattern/launch/launch_core.hpp>`_.

Camp Used in RAJAPerf
=====================

> RAJAPerf is RAJA's Performance Suite designed to explore the performance of loop-based computational kernels found in HPC applications.

Reviewer: ```suggestion
`RAJAPerf <https://github.com/LLNL/RAJAPerf>`_ is RAJA's Performance Suite designed to explore the performance of loop-based computational kernels found in HPC applications.
```

Specifically, it is used to assess and monitor runtime performance of kernels implemented using RAJA and compare those to variants 
implemented using common parallel programming models, such as OpenMP and CUDA, directly.
Camp has also been included in RAJAPerf as a way to easily determine which stream to run a RAJA kernel:

.. code-block:: cpp

   #if defined(RAJA_ENABLE_CUDA)
     camp::resources::Cuda getCudaResource()
     {
       if (run_params.getGPUStream() == 0) {
         return camp::resources::Cuda::CudaFromStream(0);
       }
       return camp::resources::Cuda::get_default();
     }
   #endif

See the full example `here <https://github.com/LLNL/RAJAPerf/blob/abb07792a899f7417e77ea40015e7e1dfd52716e/src/common/KernelBase.hpp>`_.

Camp Used in CHAI
=================

> CHAI is a library that handles automatic data migration to different memory spaces behind an array-style interface. It was designed to 

Reviewer: ```suggestion
`CHAI <https://github.com/LLNL/CHAI>`_ is a library that handles automatic data migration to different memory spaces behind an array-style interface. It was designed to 
```

.. code-block:: cpp

   static void copy(void * dst_pointer, void * src_pointer, umpire::ResourceManager & manager, ExecutionSpace dst_space, ExecutionSpace src_space) {

   #ifdef CHAI_ENABLE_CUDA
     camp::resources::Resource device_resource(camp::resources::Cuda::get_default());
   #elif defined(CHAI_ENABLE_HIP)
     camp::resources::Resource device_resource(camp::resources::Hip::get_default());
   #else
     camp::resources::Resource device_resource(camp::resources::Host::get_default());
   #endif

     camp::resources::Resource host_resource(camp::resources::Host::get_default());
     if (dst_space == GPU || src_space == GPU) {
       // Do the copy using the device resource
       manager.copy(dst_pointer, src_pointer, device_resource);
     } else {
       // Do the copy using the host resource
       manager.copy(dst_pointer, src_pointer, host_resource);
     }
     // Ensure device to host copies are synchronous
     if (dst_space == CPU && src_space == GPU) {
       device_resource.wait();
     }
   }

See the full example `here <https://github.com/LLNL/CHAI/blob/7ba2ba89071bf836071079929af7419da475ba27/src/chai/ArrayManager.cpp#L246>`_.

> Many codes at LLNL use the different libraries within the RAJA Portability Suite. Camp plays a vital role

Reviewer: ```suggestion
Many codes at LLNL and elsewhere use the different libraries within the RAJA Portability Suite. Camp plays a vital role
```

> in the compiler abstractions that make the RAJA Portability Suite possible.

Reviewer: ```suggestion
in the software abstractions that make the RAJA Portability Suite possible.
```
